In [23]:
import pandas as pd
import numpy as np
data=pd.read_csv('C:/Users/vishn/Documents/GitHub/csc478-project/hs_data_all_fields.csv', index_col=0)
data.drop(labels=['Freshmen_On_Track_CPS_Pct_Year_2','Graduation_4_Year_School_Pct_Year_2','Graduation_4_Year_School_Pct_Year_1','Graduation_5_Year_School_Pct_Year_1','College_Enrollment_CPS_Pct_Year_2','College_Enrollment_School_Pct_Year_1','College_Persistence_CPS_Pct_Year_2','Mobility_Rate_Pct','College_Persistence_School_Pct_Year_1','College_Persistence_CPS_Pct_Year_1','College_Enrollment_CPS_Pct_Year_1','Graduation_5_Year_CPS_Pct_Year_1','Graduation_5_Year_CPS_Pct_Year_2','Graduation_4_Year_CPS_Pct_Year_1','Graduation_4_Year_CPS_Pct_Year_2','Freshmen_On_Track_School_Pct_Year_1','Freshmen_On_Track_CPS_Pct_Year_1','Unnamed: 0.1','School_Hours','Legacy_Unit_ID','Finance_ID','Short_Name','Summary','Primary_Category','Is_High_School','Is_Middle_School','Is_Elementary_School','Is_Pre_School','Administrator_Title','Administrator','Phone','Secondary_Contact_Title','Secondary_Contact','Address','City','State','Zip','Fax','Freshman_Start_End_Time','After_School_Hours','Earliest_Drop_Off_Time','Significantly_Modified','Blue_Ribbon_Award_Year','Spot_Light_Award_Year','Improvement_Award_Year','Excellence_Award_Year','Hard_Of_Hearing','Mean_ACT','College_Enrollment_Rate_Mean','School_Year','Progress_Report_Year','Graduation_Rate_Mean','Visual_Impairments','School_Latitude','Classroom_Languages','CPS_School_Profile','Website','Facebook','Twitter','Youtube','Pinterest','Attendance_Boundaries','Grades_Offered_All','Grades_Offered','Student_Count_Native_American','Student_Count_Other_Ethnicity','Student_Count_Asian_Pacific_Islander','Student_Count_Multi','Student_Count_Hawaiian_Pacific_Islander','Student_Count_Ethnicity_Not_Available','Statistics_Description','Demographic_Description','PreK_School_Day','Kindergarten_School_Day','PreSchool_Inclusive','Preschool_Instructional','School_Longitude','School_Longitude','Rating_Statement','Classification_Description','Third_Contact_Title','Third_Contact_Name','Fourth_Contact_Title','Fourth_Contact_Name','Fifth_Contact_Title','Fifth_Contact_Name','Sixth_Contact_Title','Sixth_Contact_Name','Seventh_Contact_Title','Seventh_Contact_Name','Location','Excelerate_Award_Gold_Year','Student_Growth_Description','Growth_Reading_Grades_Tested_Pct_ES','Growth_Reading_Grades_Tested_Label_ES','Growth_Math_Grades_Tested_Pct_ES','Growth_Math_Grades_Tested_Label_ES','Student_Attainment_Description','Attainment_Reading_Pct_ES','Attainment_Reading_Lbl_ES','Attainment_Math_Pct_ES','Attainment_Math_Lbl_ES','Culture_Climate_Description','School_Survey_Student_Response_Rate_Avg_Pct','School_Survey_Teacher_Response_Rate_Avg_Pct','Healthy_School_Certification','Healthy_School_Certification_Description','School_Survey_Student_Response_Rate_Pct','School_Survey_Teacher_Response_Rate_Pct','Creative_School_Certification','Creative_School_Certification_Description','NWEA_Reading_Growth_Grade_3_Pct','NWEA_Reading_Growth_Grade_3_Lbl','NWEA_Reading_Growth_Grade_4_Pct','NWEA_Reading_Growth_Grade_4_Lbl','NWEA_Reading_Growth_Grade_5_Pct','NWEA_Reading_Growth_Grade_5_Lbl','NWEA_Reading_Growth_Grade_6_Pct','NWEA_Reading_Growth_Grade_6_Lbl','NWEA_Reading_Growth_Grade_7_Pct','NWEA_Reading_Growth_Grade_7_Lbl','NWEA_Reading_Growth_Grade_8_Pct','NWEA_Reading_Growth_Grade_8_Lbl','NWEA_Math_Growth_Grade_3_Pct','NWEA_Math_Growth_Grade_3_Lbl','NWEA_Math_Growth_Grade_4_Pct','NWEA_Math_Growth_Grade_4_Lbl','NWEA_Math_Growth_Grade_5_Pct','NWEA_Math_Growth_Grade_5_Lbl','NWEA_Math_Growth_Grade_6_Pct','NWEA_Math_Growth_Grade_6_Lbl','NWEA_Math_Growth_Grade_7_Pct','NWEA_Math_Growth_Grade_7_Lbl','NWEA_Math_Growth_Grade_8_Pct','NWEA_Math_Growth_Grade_8_Lbl','NWEA_Reading_Attainment_Grade_2_Pct','NWEA_Reading_Attainment_Grade_2_Lbl','NWEA_Reading_Attainment_Grade_3_Pct','NWEA_Reading_Attainment_Grade_3_Lbl','NWEA_Reading_Attainment_Grade_4_Pct','NWEA_Reading_Attainment_Grade_4_Lbl','NWEA_Reading_Attainment_Grade_5_Pct','NWEA_Reading_Attainment_Grade_5_Lbl','NWEA_Reading_Attainment_Grade_6_Pct','NWEA_Reading_Attainment_Grade_6_Lbl','NWEA_Reading_Attainment_Grade_7_Pct','NWEA_Reading_Attainment_Grade_7_Lbl','NWEA_Reading_Attainment_Grade_8_Pct','NWEA_Reading_Attainment_Grade_8_Lbl','NWEA_Math_Attainment_Grade_2_Pct','NWEA_Math_Attainment_Grade_2_Lbl','NWEA_Math_Attainment_Grade_3_Pct','NWEA_Math_Attainment_Grade_3_Lbl','NWEA_Math_Attainment_Grade_4_Pct','NWEA_Math_Attainment_Grade_4_Lbl','NWEA_Math_Attainment_Grade_5_Pct','NWEA_Math_Attainment_Grade_5_Lbl','NWEA_Math_Attainment_Grade_6_Pct','NWEA_Math_Attainment_Grade_6_Lbl','NWEA_Math_Attainment_Grade_7_Pct','NWEA_Math_Attainment_Grade_7_Lbl','NWEA_Math_Attainment_Grade_8_Pct','NWEA_Math_Attainment_Grade_8_Lbl','Suspensions_Per_100_Students_Avg_Pct','Misconducts_To_Suspensions_Avg_Pct','Average_Length_Suspension_Year_1_Pct','Average_Length_Suspension_Year_2_Pct','Average_Length_Suspension_Avg_Pct','Behavior_Discipline_Year_1','Behavior_Discipline_Year_2','Student_Attendance_Avg_Pct','Teacher_Attendance_Avg_Pct','One_Year_Dropout_Rate_Avg_Pct','Other_Metrics_Year_1','Other_Metrics_Year_2','Growth_ACT_Grade_11_Lbl','Attainment_ACT_Grade_11_Lbl','Progress_Toward_Graduation_Year_1','Progress_Toward_Graduation_Year_2','State_School_Report_Card_URL','Chronic_Truancy_Pct','Empty_Progress_Report_Message','School_Survey_Rating_Description','Supportive_School_Award_Desc','Parent_Survey_Results_Year','School_Latitude.1','School_Longitude.1','Transportation_Bus','Transportation_El','Transportation_Metra'], axis=1,inplace=True)
data.set_index('School_ID', inplace=True)
name=data['Long_Name']
data.drop(labels='Long_Name', axis=1, inplace=True)
data=data[-data['Graduation_Rate_School'].isna()]
data.describe(include='all')

School_Type  Student_Count_Total  Student_Count_Low_Income  \
count            121           121.000000                121.000000   
unique            10                  NaN                       NaN   
top     Neighborhood                  NaN                       NaN   
freq              44                  NaN                       NaN   
mean             NaN           768.181818                617.363636   
std              NaN           684.097033                474.683105   
min              NaN            38.000000                 37.000000   
25%              NaN           307.000000                271.000000   
50%              NaN           556.000000                479.000000   
75%              NaN          1062.000000                855.000000   
max              NaN          4447.000000               2713.000000   

        Student_Count_Special_Ed  Student_Count_English_Learners  \
count                 121.000000                      121.000000   
unique                       NaN                             NaN   
top                          NaN                             NaN   
freq                         NaN                             NaN   
mean                  114.173554                       63.553719   
std                    77.856993                       93.828207   
min                     6.000000                        0.000000   
25%                    58.000000                        4.000000   
50%                    89.000000                       20.000000   
75%                   161.000000                       80.000000   
max                   394.000000                      502.000000   

        Student_Count_Black  Student_Count_Hispanic  Student_Count_White  \
count            121.000000              121.000000           121.000000   
unique                  NaN                     NaN                  NaN   
top                     NaN                     NaN                  NaN   
freq                    NaN                     NaN                  NaN   
mean             284.090909              353.603306            77.528926   
std              287.302628              463.267174           227.987320   
min                1.000000                1.000000             0.000000   
25%               90.000000               13.000000             1.000000   
50%              196.000000              145.000000             4.000000   
75%              355.000000              539.000000            34.000000   
max             1468.000000             2377.000000          1567.000000   

        Student_Count_Asian        ADA_Accessible Dress_Code  \
count            121.000000                   121        121   
unique                  NaN                     3          2   
top                     NaN  Generally accessible          Y   
freq                    NaN                    66         78   
mean              35.595041                   NaN        NaN   
std               86.574205                   NaN        NaN   
min                0.000000                   NaN        NaN   
25%                0.000000                   NaN        NaN   
50%                1.000000                   NaN        NaN   
75%               10.000000                   NaN        NaN   
max              469.000000                   NaN        NaN   

       Bilingual_Services Refugee_Services Title_1_Eligible  \
count                 115              113              121   
unique                  2                2                2   
top                     Y                N                Y   
freq                   67              107              111   
mean                  NaN              NaN              NaN   
std                   NaN              NaN              NaN   
min                   NaN              NaN              NaN   
25%                   NaN              NaN              NaN   
50%                   NaN              NaN              NaN   
75%                   NaN  

In the above step we are deleting irrelevant column names, subsetting the data to school that have graduation rate value filled out, separating names of schools for future use if needed, changing index to school id. 

Next is to create change values from fields that have year 1 and year 2 and drop count variables

In [24]:
data['Suspensions_Per_100_Students_Change']=data['Suspensions_Per_100_Students_Year_1_Pct']-data['Suspensions_Per_100_Students_Year_2_Pct']
data['Misconducts_To_Suspensions_Change']=data['Misconducts_To_Suspensions_Year_1_Pct']-data['Misconducts_To_Suspensions_Year_2_Pct']
data['Student_Attendance_Change']=data['Student_Attendance_Year_1_Pct']-data['Student_Attendance_Year_2_Pct']
data['Teacher_Attendance_Change']=data['Teacher_Attendance_Year_1_Pct']-data['Teacher_Attendance_Year_2_Pct']
data['One_Year_Dropout_Rate_Change']=data['One_Year_Dropout_Rate_Year_1_Pct']-data['One_Year_Dropout_Rate_Year_2_Pct']
data.drop(labels=['College_Enrollment_School_Pct_Year_2','Graduation_5_Year_School_Pct_Year_2','Suspensions_Per_100_Students_Year_1_Pct','Suspensions_Per_100_Students_Year_2_Pct','Misconducts_To_Suspensions_Year_1_Pct','Misconducts_To_Suspensions_Year_2_Pct','Student_Attendance_Year_1_Pct','Student_Attendance_Year_2_Pct','Teacher_Attendance_Year_1_Pct','Teacher_Attendance_Year_2_Pct','One_Year_Dropout_Rate_Year_1_Pct','One_Year_Dropout_Rate_Year_2_Pct'],axis=1,inplace=True)

Following step fills in null values:

average for missing values of ACT Score, college enrollment rate, and other percent values related to school or college performance

zeroes for rates of change from previous year; we are assuming that schools that had missing values for suspension, misconduct, teacher attendance and one year drop out rates, had no changces from 2015 to 2016

U for unknown for categorical values that otherwise have a Y/N indicator

In [25]:
data['Average_ACT_School'].fillna(data['Average_ACT_School'].mean(),inplace=True)
data['College_Enrollment_Rate_School'].fillna(data['College_Enrollment_Rate_School'].mean(),inplace=True)
data['Growth_ACT_Grade_11_Pct'].fillna(data['Growth_ACT_Grade_11_Pct'].mean(),inplace=True)
data['Attainment_ACT_Grade_11_Pct'].fillna(data['Attainment_ACT_Grade_11_Pct'].mean(),inplace=True)
data['Freshmen_On_Track_School_Pct_Year_2'].fillna(data['Freshmen_On_Track_School_Pct_Year_2'].mean(),inplace=True)
data['College_Persistence_School_Pct_Year_2'].fillna(data['College_Persistence_School_Pct_Year_2'].mean(),inplace=True)

data['Suspensions_Per_100_Students_Change'].fillna(0,inplace=True)
data['Misconducts_To_Suspensions_Change'].fillna(0,inplace=True)
data['Teacher_Attendance_Change'].fillna(0,inplace=True)
data['One_Year_Dropout_Rate_Change'].fillna(0,inplace=True)

data['Bilingual_Services'].fillna('U',inplace=True)
data['Refugee_Services'].fillna('U',inplace=True)


Some values in the data set had counts of stuned in the school that were part of a group. We calculated rates for those groups using provided counts and student total count. Then, dropped the original count columns.

In [26]:
data['Student_Low_Income_Pct']=data['Student_Count_Low_Income']/data['Student_Count_Total']
data['Student_Special_Ed_Pct']=data['Student_Count_Special_Ed']/data['Student_Count_Total']
data['Student_English_Learners_Pct']=data['Student_Count_English_Learners']/data['Student_Count_Total']
data['Student_Black_Pct']=data['Student_Count_Black']/data['Student_Count_Total']
data['Student_Hispanic_Pct']=data['Student_Count_Hispanic']/data['Student_Count_Total']
data['Student_White_Pct']=data['Student_Count_White']/data['Student_Count_Total']
data['Student_Asian_Pct']=data['Student_Count_Asian']/data['Student_Count_Total']

data.drop(labels=['Student_Count_Low_Income','Student_Count_Special_Ed','Student_Count_English_Learners','Student_Count_Black','Student_Count_Hispanic','Student_Count_White','Student_Count_Asian'], axis=1,inplace=True)

In the following step we are creating a new dataset with dummies for categorical variables

In [27]:
data_dummies=pd.get_dummies(data)

Next step is to create another dataset with all variables scaled from 0 to 1. This data set will be used for regression modeling.

In [28]:
from sklearn.preprocessing import MinMaxScaler
min_max=MinMaxScaler()
min_max.fit(data_dummies)
data_scaled=pd.DataFrame(min_max.transform(data_dummies),index=data_dummies.index.values,columns=data_dummies.columns.values)

pd.set_option('display.max_columns', 100)
data_scaled.describe(include='all')

Student_Count_Total  Average_ACT_School  \
count           121.000000          121.000000   
mean              0.165612            0.264698   
std               0.155159            0.187867   
min               0.000000            0.000000   
25%               0.061012            0.127273   
50%               0.117487            0.230303   
75%               0.232252            0.327273   
max               1.000000            1.000000   

       College_Enrollment_Rate_School  Graduation_Rate_School  \
count                      121.000000              121.000000   
mean                         0.626671                0.745314   
std                          0.202823                0.198001   
min                          0.000000                0.000000   
25%                          0.489293                0.694274   
50%                          0.635974                0.780164   
75%                          0.798715                0.871166   
max                          1.000000                1.000000   

       Growth_ACT_Grade_11_Pct  Attainment_ACT_Grade_11_Pct  \
count               121.000000                   121.000000   
mean                  0.363839                     0.277600   
std                   0.206636                     0.255489   
min                   0.000000                     0.000000   
25%                   0.223881                     0.083333   
50%                   0.328358                     0.208333   
75%                   0.462687                     0.375000   
max                   1.000000                     1.000000   

       Freshmen_On_Track_School_Pct_Year_2  \
count                           121.000000   
mean                              0.745342   
std                               0.191599   
min                               0.000000   
25%                               0.670000   
50%                               0.776667   
75%                               0.886667   
max                               1.000000   

       College_Persistence_School_Pct_Year_2     lib_cnt  \
count                             121.000000  121.000000   
mean                                0.534013    0.390684   
std                                 0.221170    0.197660   
min                                 0.000000    0.000000   
25%                                 0.442478    0.272727   
50%                                 0.547198    0.363636   
75%                                 0.694690    0.454545   
max                                 1.000000    1.000000   

       PRED_NON_ENG_PERCENT  PERCENT OF HOUSING CROWDED  \
count            121.000000                  121.000000   
mean               0.255280                    0.344282   
std                0.324058                    0.277461   
min                0.000000                    0.000000   
25%                0.022222                    0.141935   
50%                0.088889                    0.232258   
75%                0.400000                    0.458065   
max                1.000000                    1.000000   

       PERCENT HOUSEHOLDS BELOW POVERTY  PERCENT AGED 16+ UNEMPLOYED  \
count                        121.000000                   121.000000   
mean                           0.388523                     0.361173   
std                            0.199614                     0.230247   
min                            0.000000                     0.000000   
25%                            0.212806                     0.169872   
50%                            0.344633                     0.355769   
75%                            0.497175                     0.516026   
max                            1.000000                     1.000000   

       PERCENT AGED 25+ WITHOUT HIGH SCHOOL DIPLOMA  \
count                                    121.000000   
mean                                       0.378996   
std                                        0.259615   
min                                        0.

Checking description of all variables I can make sure that all variable have a count of 121, with min of 0 and max of 1.

Last three steps export three datasets for use in analysis and modeling.

In [29]:
data_dummies.to_csv('processed_data.csv')

In [30]:
data.to_csv('processed_nodummies.csv')

In [31]:
data_scaled.to_csv('processed_scaled.csv')